# Day 3: Data Cleaning

## Objective

The goal of this notebook is to clean the Amazon Sales dataset by:
- Removing unnecessary columns
- Investigating missing values
- Removing duplicate records
- Converting data types
- Preparing the dataset for exploratory data analysis (EDA)

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/content/Amazon Sale Report.csv')

In [3]:
df.shape

(128976, 21)

In [4]:
df_clean = df.copy()

## Creating a Working Copy

A copy of the original dataset is created so that the raw dataset remains unchanged throughout the cleaning process.

In [5]:
df_clean[['New', 'PendingS']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128976 entries, 0 to 128975
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   New       0 non-null      float64
 1   PendingS  0 non-null      float64
dtypes: float64(2)
memory usage: 2.0 MB


In [6]:
df_clean.drop(columns=['New', 'PendingS'], inplace=True)

In [7]:
df_clean.shape

(128976, 19)

## Removing Empty Columns

The columns `New` and `PendingS` contain 100% missing values and do not provide any useful information for analysis.

These columns have been removed.

In [8]:
df_clean.duplicated().sum()

np.int64(168)

In [9]:
df_clean.drop_duplicates(inplace=True)

In [10]:
df_clean.duplicated().sum()

np.int64(0)

## Removing Duplicate Records

Duplicate records can distort analysis results.

A total of 168 duplicate rows were identified and removed from the dataset.

In [11]:
df_clean['Date'].dtype

dtype('O')

In [12]:
df_clean['Date'] = pd.to_datetime(df_clean['Date'])

/tmp/ipykernel_1019/3213019246.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_clean['Date'] = pd.to_datetime(df_clean['Date'])


In [13]:
df_clean['Date'].dtype

dtype('<M8[ns]')

## Converting Date Datatype

The Date column was originally stored as an object datatype.

It has been converted to datetime format to enable time-based analysis such as monthly sales trends and seasonal analysis.

In [14]:
df_clean[df_clean['Amount'].isnull()]['Status'].value_counts()

,count
Status,
Cancelled,7560
Shipped,208
Shipped - Delivered to Buyer,8
Shipping,8
Shipped - Returned to Seller,3
Pending,2


In [15]:
df_clean[df_clean['currency'].isnull()]['Status'].value_counts()

,count
Status,
Cancelled,7560
Shipped,208
Shipped - Delivered to Buyer,8
Shipping,8
Shipped - Returned to Seller,3
Pending,2


In [16]:
df_clean['fulfilled-by'].value_counts(dropna=False)

,count
fulfilled-by,
NaN,89595
Easy Ship,39213


In [17]:
df_clean[['ship-city',
          'ship-state',
          'ship-postal-code',
          'ship-country']].isnull().sum()

,0
ship-city,33
ship-state,33
ship-postal-code,33
ship-country,33


In [18]:
df_clean.dropna(
    subset=[
        'ship-city',
        'ship-state',
        'ship-postal-code',
        'ship-country'
    ],
    inplace=True
)

# Data Cleaning Summary

## Cleaning Actions Performed

1. Removed fully empty columns:
   - New
   - PendingS

2. Removed duplicate records.

3. Converted Date column to datetime format.

4. Investigated missing values in Amount and Currency.
   - Most missing values belong to cancelled orders.
   - These records were retained because they provide business insights.

5. Investigated fulfilled-by column.
   - Missing values appear to represent orders not fulfilled through Easy Ship.
   - Column retained for future analysis.

6. Removed rows with missing shipping information because they represented a very small portion of the dataset.

## Result

The dataset has been cleaned and prepared for exploratory data analysis (EDA).